File: sdk/python/failproofai_sdk/_runtime.py
"""The process-wide writer and event namespace.

These used to be constructed in `failproofai_sdk/__init__.py`. They live in a
leaf module now so that `_scopes` and the framework adapters can reach the
namespace without importing the `failproofai_sdk` package itself, which would be
a circular import.

Constructing `EventWriter` starts a daemon thread and registers the module-level
atexit flush, and that still happens at `import failproofai_sdk` time —
`__init__.py` imports this module, so the timing is unchanged.

Reach the namespace as `_runtime.event`, an attribute lookup at call time rather
than a `from ... import event` binding, so a test can swap in a recording
namespace and the scopes pick it up.
"""

from failproofai_sdk._events import EventNamespace
from failproofai_sdk._writer import EventWriter

writer = EventWriter()
event = EventNamespace(writer)

File: sdk/typescript/src/runtime.ts
/**
 * The process-wide writer and event namespace.
 *
 * These live in a leaf module so that `scopes.ts` and the framework adapters
 * can reach the namespace without importing the package index, which would be
 * a circular import.
 *
 * Constructing `EventWriter` starts the flush interval and registers the
 * module-level exit flush, and that happens at `import "@failproofai/sdk"`
 * time — the index imports this module, so the timing matches the Python SDK's.
 * The interval is `unref`'d, so importing this package never keeps a process
 * alive on its own.
 *
 * Reach the namespace through the `runtime` object rather than a direct named
 * import, so a test can swap in a recording namespace and the scopes and
 * adapters pick it up. A `let` export would work under ESM live bindings and
 * silently not under the CJS build, which is exactly the kind of difference
 * that shows up only in somebody else's project.
 */

import { EventNamespace } from "./events.js";
import { EventWriter } from "./writer.js";

const writer = new EventWriter();

export const runtime: { writer: EventWriter; event: EventNamespace } = {
  writer,
  event: new EventNamespace(writer),
};
